src/expression.js

```diff
@@ -607,25 +607,29 @@ pp.parseFunctionBody = function(node, isArrowFunction) {
   // If this is a strict mode function, verify that argument names
   // are not repeated, and it does not try to bind the words `eval`
   // or `arguments`.
-  if (this.strict || !isExpression && node.body.body.length && this.isUseStrict(node.body.body[0])) {
+  let refUseStrict = (!isExpression && node.body.body.length && this.isUseStrict(node.body.body[0])) ? node.body.body[0] : null;
+  if (this.strict || refUseStrict) {
     let oldStrict = this.strict
     this.strict = true
     if (node.id)
       this.checkLVal(node.id, true)
-    this.checkParams(node)
+    this.checkParams(node, refUseStrict)
     this.strict = oldStrict
   } else if (isArrowFunction) {
-    this.checkParams(node)
+    this.checkParams(node, refUseStrict)
   }
 }
 
 // Checks function params for various disallowed patterns such as using "eval"
 // or "arguments" and duplicate parameters.
 
-pp.checkParams = function(node) {
+pp.checkParams = function(node, refUseStrict) {
     let nameHash = {}
-    for (let i = 0; i < node.params.length; i++)
+    for (let i = 0; i < node.params.length; i++) {
+      if (node.params.type !== "Identifier" && refUseStrict && this.options.ecmaVersion >= 7)
+        this.raiseRecoverable(refUseStrict.start, "Illegal 'use strict' directive in function with non-simple parameter list");
```

Comment:
How about inverting an error? (erroring on non-simple parameter in function with `use strict`) In that case you wouldn't need to pass around that reference. Or do you think it would be less clear?